netdog: Enable conditional compilation based on network backend

zmrow opened the pull request:
**Issue number:**
Related to #2449 

**Description of changes:**
This set of temporary changes enables conditional compilation of `netdog` based on network backend.  "Temporary" in this case means that these changes (similar to those in #2741 ) will be removed once `systemd-networkd` development has finished.

A new function `emit_net_backend_cfgs()` has been added to `netdog`'s `build.rs`  that triggers off the `SYSTEMD_NETWORKD` environment variable that has been added to the build environment.  The crate can now use the following to conditionally compile code:
```
#[cfg(net_backend = "systemd-networkd")]
{ do stuff }
```

**Testing done:**
Added the following blurb to `netdog generate-net-config` to be conditionally compiled:
```
#[cfg(net_backend = "systemd-networkd")]
{
    eprintln!("I'm on systemd-networkd!!!")
}
```

* Build and boot an aws-k8s-1.22 image that runs properly and connects to the cluster - do not observe the above lines in the journal
* Build and locally boot an aws-dev variant. I added an early dependency on console-getty.service to the preconfigured.target since I knew early services would be broken but I still wanted to be able to poke around and verify the above. With the `systemd-networkd` build flag, I observed the above message in the journal.  Without the flag, the message didn't show up.
```
Starting Network Configuration...
[    2.081577] netdog[901]: I'm on systemd-networkd!!!
```



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
Dockerfile: Set SYSTEMD_NETWORKD environment variable in build stage

This commit sets the `SYSTEMD_NETWORKD` environment variable during the
RPM build stage, which is useful to trigger on and emit additional Rust
cfgs for conditional compilation.

bcressey commented on sources/bottlerocket-variant/src/lib.rs:
```diff
@@ -370,6 +381,20 @@ impl PartialEq<Variant> for &str {
     }
 }
 
+/// This can be used in a `build.rs` file to emit `cfg` values that can be used for conditional
+/// compilation based on the network backend. This function also emits rerun-if-changed so
+/// that packages sensitive to network backend will rebuild if the backend changes.
+/// TODO: Remove this function once systemd-networkd integration development is finished
+pub fn emit_net_backend_cfgs() {
+    let net_backend = if Some("1".to_string()) == std::env::var(NET_BACKEND_OVERRIDE_ENV).ok() {
+        NET_BACKEND_OVERRIDE
+    } else {
+        DEFAULT_NET_BACKEND
+    };
+    println!("cargo:rerun-if-env-changed={}", NET_BACKEND_OVERRIDE_ENV);
+    println!("cargo:rustc-cfg=net_backend=\"{}\"", net_backend);
+}
```
I'd rather do this in a netdog-specific build script. It's not something most other crates will care about, while quite a few benefit from structured parsing of the variant tuple.

webern replied:
> I'd rather do this in a netdog-specific build script. It's not something most other crates will care about, while quite a few benefit from structured parsing of the variant tuple.

Not sure why I didn't think that. I think when we discussed it, @zmrow, it was only netdog that needed this?

zmrow replied:
That's fair.  My reasoning for putting it in the variant crate was "just in case we need it outside of netdog".  I'm fine with scoping it to netdog for now.

bcressey reviewed (commented):
I'm on the fence about merging this in the current state, since there's no real functionality tied to it, and it seems like an easy patch to maintain on a feature branch. If you can scope it down to `netdog` then it's probably OK.

zmrow committed:
netdog: Enable conditional compilation based on network backend

This commit enables conditional compilation of `netdog` based on the
network backend.  A temporary function has been added that reads the
environment variable `SYSTEMD_NETWORKD`, and emits additional cfgs to
direct compilation.  This environment variable will be passed to the
build if a user has enabled the build flag to use systemd-networkd
rather than wicked in their variant.

zmrow commented:
^ Scoped these changes to `netdog` per @bcressey

webern commented on sources/api/netdog/build.rs:
```diff
@@ -14,6 +25,19 @@ fn main() {
+/// TODO: Remove this function once systemd-networkd integration development is finished
+fn emit_net_backend_cfgs() {
+    let net_backend = if Some("1".to_string()) == std::env::var(NET_BACKEND_OVERRIDE_ENV).ok() {
+        NET_BACKEND_OVERRIDE
+    } else {
+        DEFAULT_NET_BACKEND
```
very much a nit: our other `cfg!` compile guards use lowercase, kebab-case strings, e.g. `aws-k8s` etc. So these screaming snake cases might look a little out of place at their usage sites.

zmrow replied:
The actual cfgs emitted will be: `wicked` and `systemd-networkd`.  These screaming snake case constants are set to those static strings. :)

webern replied:
ah i see :+1:
